### DISCUSSION TITLE
data formatting issue while running prediction from AR model

### DISCUSSION THREAD

**Discussion by barbeyj7** [community/community#118336] [Programming Help] [answered]
### Body

Hello, I am struggling finding the solution for the following error while running the below code
 "TypeError: float() argument must be a string or a real number, not 'Period'"

-> I tried to change the index with pd.DatetimeIndex or to_timestamp but was unsuccessful.

would appreciate any idea

thanks in advance !

----

import numpy as np
import pandas as pd
import matplotlib.pyplot as plt
import datetime as dt
import pandas_datareader as pdr
import statsmodels.api as sm
from matplotlib.pyplot import plot, scatter, show, xlabel, ylabel
import yfinance as yf
from statsmodels.tsa.ar_model import AutoReg
from datetime import datetime
from datetime import timedelta

start = '2021-01-01'
end = "2024-03-22"

stock1 = yf.Ticker('AAPL')   
stock1_data = stock1.history(interval='1d', start= start, end= end)

stock2 = yf.Ticker('AMZN')   
stock2_data = stock2.history(interval='1d', start= start, end= end)

X = np.log(stock1_data["Close"])

Y = np.log(stock2_data["Close"])

X = sm.add_constant(X)

model = sm.OLS(Y,X)

results = model.fit()

alpha = results.params.values[0]

beta = results.params.values[1]

errors = Y - (alpha + X["Close"]*beta)

errors.index = pd.DatetimeIndex(errors.index).to_period('D')
train_end = datetime(2023,4,1)

test_end = datetime(2024,4,1)

train_data = errors[:train_end]

test_data = errors[train_end+timedelta(days=1):test_end]

model2 = AutoReg(train_data, lags=3)

model2_fit = model2.fit()

pred_start_date = test_data.index[0]

pred_end_date = test_data.index[-1]

predictions = model2_fit.predict(start=pred_start_date, end=pred_end_date)

plt.plot(predictions)

### Guidelines

- [X] I have read and understood this category's [guidelines](https://github.com/orgs/community/discussions/51384) before making this post.

**Comment by barbeyj7** [comment]
also tried : 
pandas.plotting.deregister_matplotlib_converters

but not helping here

**Comment by MarouaneBenbetka** [comment]
Hello @barbeyj7  :

check this code it may work , and if it answers your question mark as answered , GOOD LUCK !
```
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt
import statsmodels.api as sm
import yfinance as yf
from statsmodels.tsa.ar_model import AutoReg
from datetime import datetime

# Define the start and end dates for fetching the stock data
start = '2021-01-01'
end = "2024-03-22"

# Fetch historical data for AAPL
stock1 = yf.Ticker('AAPL')
stock1_data = stock1.history(interval='1d', start=start, end=end)

# Fetch historical data for AMZN
stock2 = yf.Ticker('AMZN')
stock2_data = stock2.history(interval='1d', start=start, end=end)

# Calculate the log of the closing prices
X = np.log(stock1_data["Close"])
Y = np.log(stock2_data["Close"])

# Add a constant term for the linear regression model
X = sm.add_constant(X)

# Fit the linear regression model
model = sm.OLS(Y, X)
results = model.fit()

# Extract the model parameters
alpha = results.params.values[0]
beta = results.params.values[1]

# Calculate the residuals/errors
errors = Y - (alpha + X["Close"] * beta)

# Ensure the errors index is tz-naive
errors.index = errors.index.tz_localize(None)

# Define the training and testing period
train_end = datetime(2023, 4, 1)
test_end = datetime(2024, 3, 22)

# Split the data into training and testing datasets
train_data = errors[:train_end]
test_data = errors[train_end + pd.Timedelta(days=1):test_end]

# Fit the AutoRegressive model
model2 = AutoReg(train_data, lags=3)
model2_fit = model2.fit()

# Instead of using pred_start_date and pred_end_date directly, use integer locations.
# This is where the training data ends, and the prediction period starts
start_loc = train_data.shape[0]
end_loc = start_loc + test_data.shape[0] - 1

predictions = model2_fit.predict(start=start_loc, end=end_loc)

predictions = pd.Series(predictions, index=test_data.index)

plt.figure(figsize=(10, 6))
plt.plot(predictions.index, predictions, label='Predictions', color='red')
plt.plot(errors.index, errors, label='Errors', alpha=0.5, color='blue')
plt.legend()
plt.xlabel('Date')
plt.ylabel('Error / Prediction')
plt.title('Error and Predictions Over Time')
plt.show()

```

**Reply by barbeyj7** [reply]
Hello @MarouaneBenbetka 
Thanks a lot for the feedback and improvements ! 
Was not aware of tz.localize function + didn't think of reindexing through pd.Series.

I am almost there, however I can not see the predictions in the plot
![Capture d’écran 2024-04-09 à 15 37 47](https://github.com/community/community/assets/141441563/ca72e385-3d5f-4227-bae2-56ffccce2160)

My aim is to plot the actual data (errors) vs. the predicted one (via training) as in the attached example.

btw, why can't I simply use plt.plot(predictions) as you previously defined the index ?

Thank you

**Comment by jfullstackdev** [comment]
from Copilot suggestion : 

---

The error you're encountering is due to the fact that the `predict` method of the `AutoReg` model from `statsmodels` doesn't support `Period` index, which you're trying to use after converting your `DatetimeIndex` to `Period` index.

You should keep your index as `DatetimeIndex` for the prediction. Here's the corrected part of your code:

```python
# Don't convert the index to Period
# errors.index = pd.DatetimeIndex(errors.index).to_period('D')

train_end = datetime(2023,4,1)
test_end = datetime(2024,4,1)

# Use loc to slice the DataFrame by date
train_data = errors.loc[:train_end]
test_data = errors.loc[train_end+timedelta(days=1):test_end]

model2 = AutoReg(train_data, lags=3)
model2_fit = model2.fit()

pred_start_date = test_data.index[0]
pred_end_date = test_data.index[-1]

predictions = model2_fit.predict(start=pred_start_date, end=pred_end_date)

plt.plot(predictions)
```

This should resolve the `TypeError` you're encountering.

**Reply by barbeyj7** [reply]
Hello @jdevstatic 
actually I still have an error : "TypeError: Cannot compare tz-naive and tz-aware datetime-like objects"
I assume there might be a proper way to clean the errors index.

**Reply by jfullstackdev** [reply]
can you give me now your updated code ? so I can see it all,

**Reply by barbeyj7** [reply]
sure, hereafter : 

```
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt
import statsmodels.api as sm
import yfinance as yf
from statsmodels.tsa.ar_model import AutoReg
from datetime import datetime
from datetime import timedelta

# Define the start and end dates for fetching the stock data
start = '2021-01-01'
end = "2024-03-22"

# Fetch historical data for AAPL
stock1 = yf.Ticker('AAPL')
stock1_data = stock1.history(interval='1d', start=start, end=end)

# Fetch historical data for AMZN
stock2 = yf.Ticker('AMZN')
stock2_data = stock2.history(interval='1d', start=start, end=end)

# Calculate the log of the closing prices
X = np.log(stock1_data["Close"])
Y = np.log(stock2_data["Close"])

# Add a constant term for the linear regression model
X = sm.add_constant(X)

# Fit the linear regression model
model = sm.OLS(Y, X)
results = model.fit()

# Extract the model parameters
alpha = results.params.values[0]
beta = results.params.values[1]

# Calculate the residuals/errors
errors = Y - (alpha + X["Close"] * beta)
errors.index

# Define the training and testing period
train_end = datetime(2023, 4, 1)
test_end = datetime(2024, 3, 22)

# Use loc to slice the DataFrame by date
train_data = errors.loc[:train_end]
test_data = errors.loc[train_end + timedelta(days=1):test_end]

# Fit the AutoRegressive model
model2 = AutoReg(train_data, lags=3)
model2_fit = model2.fit()

pred_start_date = test_data.index[0]
pred_end_date = test_data.index[-1]

predictions = model2_fit.predict(start=pred_start_date, end=pred_end_date)

plt.plot(predictions)
```

**Reply by jfullstackdev** [reply]
The error message "TypeError: Cannot compare tz-naive and tz-aware datetime-like objects" suggests that you're trying to compare or use datetime objects that are timezone aware with those that are not. 

In your case, the error is likely occurring when you're slicing your DataFrame by date. The `errors` DataFrame index (dates) is timezone naive (it doesn't have timezone information), while `train_end` and `test_end` are timezone aware (they have timezone information).

To fix this, you can make `train_end` and `test_end` timezone naive by not providing any timezone information when creating these datetime objects:

```python
# Define the training and testing period
train_end = datetime(2023, 4, 1)
test_end = datetime(2024, 3, 22)
```

If you still encounter the error, you can explicitly make these datetime objects timezone naive:

```python
# Define the training and testing period
train_end = datetime(2023, 4, 1).replace(tzinfo=None)
test_end = datetime(2024, 3, 22).replace(tzinfo=None)
```

This should resolve the TypeError you're encountering.

**Reply by barbeyj7** [reply]
still not unfortunately.
btw, errors.index indicates dtype='datetime64[ns, America/New_York]', timezone aware, isn't it ?

**Reply by jfullstackdev** [reply]
is there a new error ? or still the last error ?

**Reply by barbeyj7** [reply]
same error ```TypeError: Cannot compare tz-naive and tz-aware datetime-like objects```

but now I have it directly after this line of code : 
```train_data = errors.loc[:train_end]```

**Reply by jfullstackdev** [reply]
I'm just wondering how you edit the code, can I see the full code again ?

**Reply by barbeyj7** [reply]
Yes, just added ```replace``` as you suggested

```
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt
import statsmodels.api as sm
import yfinance as yf
from statsmodels.tsa.ar_model import AutoReg
from datetime import datetime
from datetime import timedelta

# Define the start and end dates for fetching the stock data
start = '2021-01-01'
end = "2024-03-22"

# Fetch historical data for AAPL
stock1 = yf.Ticker('AAPL')
stock1_data = stock1.history(interval='1d', start=start, end=end)

# Fetch historical data for AMZN
stock2 = yf.Ticker('AMZN')
stock2_data = stock2.history(interval='1d', start=start, end=end)

# Calculate the log of the closing prices
X = np.log(stock1_data["Close"])
Y = np.log(stock2_data["Close"])

# Add a constant term for the linear regression model
X = sm.add_constant(X)

# Fit the linear regression model
model = sm.OLS(Y, X)
results = model.fit()

# Extract the model parameters
alpha = results.params.values[0]
beta = results.params.values[1]

# Calculate the residuals/errors
errors = Y - (alpha + X["Close"] * beta)

# Define the training and testing period
train_end = datetime(2023, 4, 1).replace(tzinfo=None)
test_end = datetime(2024, 3, 22).replace(tzinfo=None)

# Use loc to slice the DataFrame by date
train_data = errors.loc[:train_end]
test_data = errors.loc[train_end + timedelta(days=1):test_end]

# Fit the AutoRegressive model
model2 = AutoReg(train_data, lags=3)
model2_fit = model2.fit()

pred_start_date = test_data.index[0]
pred_end_date = test_data.index[-1]

predictions = model2_fit.predict(start=pred_start_date, end=pred_end_date)

plt.plot(predictions)
```

**Reply by jfullstackdev** [reply]
maybe in this part ? though not so sure

```
# Convert the index to timezone naive datetime objects
errors.index = errors.index.tz_localize(None)

# Now this line should work without errors
train_data = errors.loc[:train_end]
```

**Reply by barbeyj7** [reply]
```
model2 = AutoReg(train_data, lags=3)
/Users/mathieudepaigne/anaconda3/lib/python3.11/site-packages/statsmodels/tsa/base/tsa_model.py:473: ValueWarning: A date index has been provided, but it has no associated frequency information and so will be ignored when e.g. forecasting.
  self._init_dates(dates, freq)
```

**Reply by jfullstackdev** [reply]
how about this ? 

```
# Add a daily frequency to the DateTime index
train_data.index = pd.DatetimeIndex(train_data.index).to_period('D')

# Now this line should work without warnings
model2 = AutoReg(train_data, lags=3)
```

**Reply by barbeyj7** [reply]
thanks,  it's ok for the AutoReg line
But still have error message when trying to plot and I don't get the point.
```TypeError: float() argument must be a string or a real number, not 'Period' ```

Only way I found to plot is by using
```predictions.asfreq('D').plot()```

**Reply by jfullstackdev** [reply]
so it's okay by now ? or you still have problems with your workaround ?

**Reply by barbeyj7** [reply]
well, can't figure out the last TypeError so if anyone has an explanation, I am a taker !

**Reply by jfullstackdev** [reply]
I can try again, can you give me the latest updated code ?

**Comment by barbeyj7** [comment]
Hi, please see below

```
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt
import statsmodels.api as sm
import yfinance as yf
from statsmodels.tsa.ar_model import AutoReg
from datetime import datetime
from datetime import timedelta

# Define the start and end dates for fetching the stock data
start = '2021-01-01'
end = "2024-04-12"

# Fetch historical data for AAPL
stock1 = yf.Ticker('AAPL')
stock1_data = stock1.history(interval='1d', start=start, end=end)

# Fetch historical data for AMZN
stock2 = yf.Ticker('AMZN')
stock2_data = stock2.history(interval='1d', start=start, end=end)

# Calculate the log of the closing prices
X = np.log(stock1_data["Close"])
Y = np.log(stock2_data["Close"])

# Add a constant term for the linear regression model
X = sm.add_constant(X)

# Fit the linear regression model
model = sm.OLS(Y, X)
results = model.fit()

# Extract the model parameters
alpha = results.params.values[0]
beta = results.params.values[1]

# Calculate the residuals/errors
errors = Y - (alpha + X["Close"] * beta)

# Convert to naive index
errors.index = errors.index.tz_localize(None)

# Define the training and testing period
train_end = datetime(2023, 4, 1)
test_end = datetime(2024, 3, 22)

# Use loc to slice the DataFrame by date
train_data = errors.loc[:train_end]
test_data = errors.loc[train_end + timedelta(days=1):test_end]

# Associate frequency to training data
train_data.index = pd.DatetimeIndex(train_data.index).to_period('D')

# Fit the AutoRegressive model
model2 = AutoReg(train_data, lags=3)
model2_fit = model2.fit()

# Run the predictions
pred_start_date = test_data.index[0]
pred_end_date = test_data.index[-1]

predictions = model2_fit.predict(start=pred_start_date, end=pred_end_date)

#Plot
plt.plot(predictions)
```

**Reply by jfullstackdev** [reply, marked as answer]
matplotlib's plot function does not support pandas Period objects, which is the type of your index after you convert it to a period with `to_period('D')`

`asfreq('D')` function converts the PeriodIndex back to a DateTimeIndex, which matplotlib can handle. That's why `predictions.asfreq('D').plot()` works.

if you want to use the standard `plt.plot()` function, you can convert the index of your predictions back to a datetime index before plotting, maybe like this

```python
predictions.index = predictions.index.to_timestamp()
plt.plot(predictions)
```

this will convert the PeriodIndex back to a DateTimeIndex, allowing you to plot the predictions with matplotlib's plot function,

**Reply by barbeyj7** [reply]
Thanks for the explanation, it works now, you rock !
Actually, when using ```errors.index = pd.DatetimeIndex(errors.index).to_period('D')``` at the very beginning, I tried to convert it back to date time object with ```errors.index = pd.DatetimeIndex(errors.index).to_timestamp()``` but it didn't work. Will give it a second thought.
